### DISCUSSION TITLE
Workflow with strict ordering

### DISCUSSION THREAD

**Discussion by abrunner94** [bensheldon/good_job#1330] [General]
I'm trying to compose a workflow, where each step in the workflow is executed serially or based on a `next` attribute.

Here is an example of my `Workflow` instance:

```json
{
    "workflow": {
        "name": "test5",
        "team_id": 1,
        "actions": [
            {
                "name": "stepName1",
                "next": "stepName2",
                "args": {
                    "someKey": "verysecret"
                }
            },
            {
                "name": "stepName2",
                "next": "stepName3",
                "args": {
                    "someResult": "{{ steps.stepName1 }}"
                }
            },
             {
                "name": "stepName3",
                "next": "end",
                "args": {
                    "someResult": "{{ steps.stepName1 }}"
                }
            }
        ]
    }
}
```

Is this possible using batches? If so, how can I ensure the order of the workflow and what would that look like?

I've tried this, but looks like the execution is not following strict ordering.

```ruby
  def execute_workflow(workflow)
    batch = GoodJob::Batch.new
    batch.description = workflow.name
    batch.enqueue do
      workflow.actions.each do |action|
        WorkJob.perform_later(workflow.id, action)
      end
    end
  end
```

```ruby
# frozen_string_literal: true

class WorkJob < ApplicationJob
  include GoodJob::ActiveJobExtensions::Batches

  def perform(workflow_id, action)
     # ...

    JobResult.create!(
      workflow_id:,
      action_name: action['name'],
      next_action_name: action['next'],
      job_id:,
      batch_id: batch.id,
      result: { test: 'abc' }
    )
  end
```

**Comment by bensheldon** [comment]
There's an example in the readme, though I am totally sympathetic if it's not easy to follow:

https://github.com/bensheldon/good_job?tab=readme-ov-file#complex-batches

The general idea of getting jobs to run serially: enqueue the jobs to the batch after the previous job(s) in the batch have run. You can do that by either adding another job to the batch _within a job in the batch_ or to use the batch callback job to enqueue additional jobs (batches are mutable in GoodJob).

**Comment by abrunner94** [comment]
This is great. The callbacks work very nicely! Thank you! 🙏
